File: 124.binary-tree-maximum-path-sum.java
// @leet start
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
  private int answer = Integer.MIN_VALUE;
  public int maxPathSum(TreeNode root) {
  	dfs(root);
    return answer;
  }
  public int dfs(TreeNode root) {
    if (root == null) {
      return 0;
    }
    int left = Math.max(0, dfs(root.left));
    int right = Math.max(0, dfs(root.right));
    answer = Math.max(answer, root.val + left + right);
    return root.val + Math.max(left, right);
  }
}
// @leet end

File: 124.binary-tree-maximum-path-sum.cpp

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left),
 * right(right) {}
 * };
 */
class Solution {
public:
    int maxPathSum(TreeNode* root) {
        int maxSum = INT_MIN;
        maxPathDown(root, maxSum);
        return maxSum;
    }
    int maxPathDown(TreeNode* node, int& maxSum) {
        if (!node) return 0;
        int left = std::max(0, maxPathDown(node->left, maxSum));
        int right = std::max(0, maxPathDown(node->right, maxSum));
        maxSum = std::max(maxSum, node->val + left + right);
        return node->val + std::max(left, right);
    }
};

